### PR TITLE
Add debugger exception to error classes

### PR DESCRIPTION
**Pull request by yl-to**
*Issue #, if available:*

*Description of changes:*
- Capture SageMaker Debugger Exceptions in training toolkit.
- fixed some naming/typo/function wrapping issues regarding previous error exception capturing code.

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md) doc
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have used the regional endpoint when creating S3 and/or STS clients (if appropriate)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-training-toolkit/blob/master/README.md)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker_training/smdataparallel.py [hunk 1]
```diff
@@ -284,18 +284,20 @@ def run(self, wait=True, capture_error=False):
         cmd = self._create_command()
         cmd.extend(super(SMDataParallelRunner, self)._create_command())
         logging_config.log_script_invocation(cmd, self._env_vars)
+
+        exception_classes_all = exception_classes + process.get_debugger_exception_classes()
```
Comment (satishpasumarthi) [inline src/sagemaker_training/smdataparallel.py:288]: can we change this to `exception_classes += process.get_debugger_exception_classes()` in order for minimal code changes?

File: src/sagemaker_training/mpi.py [hunk 2]
```diff
@@ -305,23 +305,22 @@ def run(self, wait=True, capture_error=False):
         logging_config.log_script_invocation(cmd, self._env_vars)
 
         training_env = environment.Environment()
-        exception_classes = get_modelparallel_exception_classes()
+        exception_classes = []
+        exception_classes += process.get_debugger_exception_classes()
+        if training_env.is_modelparallel_enabled:
+            exception_classes += get_modelparallel_exception_classes()
```
Comment (satishpasumarthi) [inline src/sagemaker_training/mpi.py:311]: If there are no exception classes found in sm modelparallel and debugger, it would result in `[errors.ExecuteUserScriptError, errors.ExecuteUserScriptError]`. Could you please take a look and fix this?

**Review by satishpasumarthi — CHANGES_REQUESTED** [review]
left few comments !

**Comment by satishpasumarthi** [comment]
Could you also add some testcases similar to https://github.com/satishpasumarthi/sagemaker-training-toolkit/blob/master/test/unit/test_modules.py#L236 ?

File: src/sagemaker_training/mpi.py [hunk 3]
```diff
@@ -305,23 +305,24 @@ def run(self, wait=True, capture_error=False):
...
+        if exception_classes == [errors.ExecuteUserScriptError, errors.ExecuteUserScriptError]:
```
Comment (roywei) [inline src/sagemaker_training/mpi.py:312]: what if the default exception name changed in future? what if there is another library with similar logic added? we can do a deduplication here instead of comparing to `[errors.ExecuteUserScriptError, errors.ExecuteUserScriptError]`

Comment (roywei) [reply src/sagemaker_training/mpi.py:312]: list -> set -> list

File: src/sagemaker_training/process.py [hunk 4]
```diff
@@ -39,6 +39,24 @@
...
 
+def get_debugger_exception_classes():
+    """Set exception classes"""
+    exception_classes = []
+    try:
+        from smdebug import exceptions
+
+        # list of exceptions debugger wants training toolkit to catch and log
+        exception_classes += [x for x in dir(exceptions) if isclass(getattr(exceptions, x))]
+        print(f"debug: {dir(exceptions)}")
```
Comment (roywei) [inline src/sagemaker_training/process.py:50]: remove debug print

File: src/sagemaker_training/process.py [hunk 5]
```diff
@@ -39,6 +39,24 @@
...
+    except ImportError:
+        logger.info("No exception classes found in smdebugger")
+
+    print(f"debug: {type(exception_classes[0])}")
```
Comment (roywei) [inline src/sagemaker_training/process.py:54]: remove debug print

File: src/sagemaker_training/smdataparallel.py [hunk 6]
```diff
@@ -284,6 +288,11 @@ def run(self, wait=True, capture_error=False):
         cmd = self._create_command()
         cmd.extend(super(SMDataParallelRunner, self)._create_command())
         logging_config.log_script_invocation(cmd, self._env_vars)
+
+        exception_classes = get_dataparallel_exception_classes() + \
+                            process.get_debugger_exception_classes()
+        if exception_classes == [errors.ExecuteUserScriptError, errors.ExecuteUserScriptError]:
```
Comment (roywei) [inline src/sagemaker_training/smdataparallel.py:294]: list -> set -> list to dedupe

**Review by roywei — CHANGES_REQUESTED** [review]
See comments, and please add a test case.
example: https://github.com/satishpasumarthi/sagemaker-training-toolkit/blob/master/test/unit/test_process.py#L123
also test when debugger is not enabled, or package not found does the import handles the issue.
https://github.com/satishpasumarthi/sagemaker-training-toolkit/blob/master/test/unit/test_modules.py#L236

File: src/sagemaker_training/process.py (hunk 5 above)
Comment (roywei) [reply src/sagemaker_training/process.py:54]: can you use a more detailed message like:
   "smdistributed.dataparallel not found or "
            "using an older version without custom exceptions."
            "SM training toolkit will track user script error only"

File: src/sagemaker_training/mpi.py (hunk 3 above)
Comment (yl-to) [reply src/sagemaker_training/mpi.py]: thanks! modified

File: src/sagemaker_training/process.py [hunk 7]
```diff
@@ -39,6 +39,22 @@
...
+def get_debugger_exception_classes():
+    """Set exception classes"""
+    exception_classes = []
+    try:
+        from smdebug import exceptions
+
+        # list of exceptions debugger wants training toolkit to catch and log
+        exception_classes += [ex for ex in dir(exceptions) if isclass(getattr(exceptions, ex))]
+    except ImportError:
+        logger.info("No exception classes found in smdebugger")
```
Comment (satishpasumarthi) [inline src/sagemaker_training/process.py:51]: Let's rename this to `sagemaker debugger`

Comment (satishpasumarthi) [reply src/sagemaker_training/process.py:51]: What happens when the user turns off sagemaker debugger from the estimator call ? Do we still capture exceptions from the debugger package according to this logic ? Can we differentiate between debugger disabled vs no exception classes found in debugger?

File: src/sagemaker_training/process.py [hunk 8]
```diff
@@ -40,6 +40,29 @@
 _DEFAULT_BUF_SIZE = 1024 * 64
 
 
+def get_debugger_exception_classes():
+    """Set exception classes"""
+    exception_classes = []
+    if os.environ.get("USE_SMDEBUG") == "0":
+        logger.info(
+            "sagemaker debugger is not enabled, using exception types without debugger ones."
```
Comment (satishpasumarthi) [inline src/sagemaker_training/process.py:48]: Let's rename this statement. 
`Sagemaker Debugger is not enabled, debugger exceptions will not be imported` or something better.

File: src/sagemaker_training/process.py [hunk 9]
```diff
@@ -40,6 +40,29 @@
...
+        )
+    else:
+        try:
+            from smdebug import exceptions
+
```
Comment (satishpasumarthi) [inline src/sagemaker_training/process.py:53]: remove extra line.

File: src/sagemaker_training/process.py [hunk 10]
```diff
@@ -40,6 +40,29 @@
...
+            # list of exceptions debugger wants training toolkit to catch and log
+            exception_classes += [ex for ex in dir(exceptions) if isclass(getattr(exceptions, ex))]
+        except ImportError:
+            logger.info(
+                "sagemaker debugger is not installed, using exception types without debugger ones."
```
Comment (satishpasumarthi) [inline src/sagemaker_training/process.py:58]: Please also rename this. The word `ones` should be removed from the statement. May be rephrase this to 
`Exceptions not imported for SageMaker Debugger as it is not installed`

File: src/sagemaker_training/process.py [hunk 11]
```diff
@@ -40,6 +40,29 @@
...
+            )
+
+    if not exception_classes:
+        exception_classes = [errors.ExecuteUserScriptError]
```
Comment (satishpasumarthi) [inline src/sagemaker_training/process.py:62]: Let's move this logic a bit?
At the start of the file, include `DEFAULT_ERROR_CLASS = errors.ExecuteUserScriptError` and change this to 
`exception_classes = [DEFAULT_ERROR_CLASS]`. This will be more clear to understand what is the default and how we fall back if no other classes are present.

**Review by satishpasumarthi — CHANGES_REQUESTED** [review]
Left few comments

**Comment by satishpasumarthi** [comment]
Please also fill the PR description with relevant details for easier future reference.

File: src/sagemaker_training/process.py (hunk 11 above)
Comment (yl-to) [reply src/sagemaker_training/process.py]: edited.

File: src/sagemaker_training/smdataparallel.py [hunk 12]
```diff
@@ -35,19 +35,24 @@
...
-    logger.info(
-        "smdistributed.dataparallel not found or "
-        "using an older version without custom exceptions."
-        "SM training toolkit will track user script error only"
-    )
-    exception_classes = [errors.ExecuteUserScriptError]
+
+def get_dataparallel_exception_classes():
+    """Get ddp exception classes"""
+    try:
```
Comment (roywei) [inline src/sagemaker_training/smdataparallel.py:41]: we should define  exception_classes to be an empty list first. Just to follow the convention of the other two methods.

File: src/sagemaker_training/smdataparallel.py [hunk 13]
```diff
@@ -296,6 +301,12 @@ def run(self, wait=True, capture_error=False):
         cmd = self._create_command()
         cmd.extend(super(SMDataParallelRunner, self)._create_command())
         logging_config.log_script_invocation(cmd, self._env_vars)
+
+        exception_classes = (
```
Comment (roywei) [inline src/sagemaker_training/smdataparallel.py:305]: can we keep the code style consistent? define it to a empty list and add on top of it, similar to mpi.py

**Review by roywei — APPROVED** [review]
2 minor comments, rest LGTM

**Review by roywei — APPROVED** [review]

File: src/sagemaker_training/smdataparallel.py [hunk 14]
```diff
@@ -35,19 +35,28 @@
...
+
+
+def get_dataparallel_exception_classes():
+    """Get ddp exception classes"""
+    exception_classes = []
+    try:
+        from smdistributed.dataparallel import exceptions
+
+        # list of exceptions SMDDP wants training toolkit to catch and log
+        exception_classes = [ex for ex in dir(exceptions) if isclass(getattr(exceptions, ex))]
```
Comment (roywei) [inline src/sagemaker_training/smdataparallel.py:48]: add to empty list instead of re-defining it.

File: src/sagemaker_training/smdataparallel.py [hunk 15]
```diff
@@ -296,6 +305,12 @@ def run(self, wait=True, capture_error=False):
         cmd = self._create_command()
         cmd.extend(super(SMDataParallelRunner, self)._create_command())
         logging_config.log_script_invocation(cmd, self._env_vars)
+
+        exception_classes = (
```
Comment (roywei) [inline src/sagemaker_training/smdataparallel.py:309]: I mean consistent to
```
exception_classes = []
exception_classes += process.get_debugger_exception_classes()
```

**Review by roywei — CHANGES_REQUESTED** [review]

**Review by roywei — DISMISSED** [review]

**Comment by satishpasumarthi** [comment]
Thanks for the bringing up SM Debugger exceptions into the toolkit. This PR looks good to me. 
Will approve once we decide on the plan for this release.

File: src/sagemaker_training/process.py (hunk 9 above)
Comment (yl-to) [reply src/sagemaker_training/process.py:52]: this extra line is from black check and so can't remove it.

**Review by satishpasumarthi — APPROVED** [review]
LGTM !
